pkg/services/ngalert/notifier/receiver_groups.go

```diff
@@ -0,0 +1,227 @@
+}
+
+type provisoningStore interface {
+	GetProvenances(ctx context.Context, org int64, resourceType string) (map[string]models.Provenance, error)
+}
+
+type TransactionManager interface {
+	InTransaction(ctx context.Context, work func(ctx context.Context) error) error
+}
+
+func NewReceiverGroupService(
+	ac accesscontrol.AccessControl,
+	configStore configStore,
+	provisioningStore provisoningStore,
+	encryptionService secrets.Service,
+	xact TransactionManager,
+	log log.Logger,
+) *ReceiverGroupService {
+	return &ReceiverGroupService{
+		ac:                 ac,
+		provisioningStore:  provisioningStore,
+		versionedConfStore: &LockingConfigStore{Store: configStore},
+		encryptionService:  encryptionService,
+		xact:               xact,
+		log:                log,
+	}
+}
+
+func (rs *ReceiverGroupService) canDecrypt(ctx context.Context, user identity.Requester, name string) (bool, error) {
+	receiverAccess := false // TODO: stub, check for read secrets access
+	eval := accesscontrol.EvalPermission(accesscontrol.ActionAlertingProvisioningReadSecrets)
+	provisioningAccess, err := rs.ac.Evaluate(ctx, user, eval)
+	if err != nil {
+		return false, err
+	}
+	return receiverAccess || provisioningAccess, nil
+}
+
+func (rs *ReceiverGroupService) GetReceiverGroups(ctx context.Context, q models.ReceiverGroupQuery, user identity.Requester) ([]definitions.GettableApiReceiver, error) {
+	rev, err := rs.versionedConfStore.GetLockingConfig(ctx, q.OrgID)
+	if err != nil {
+		return nil, err
+	}
+
+	provenances, err := rs.provisioningStore.GetProvenances(ctx, q.OrgID, "contactPoint")
+	if err != nil {
+		return nil, err
+	}
+
+	// TODO: check for list access
+
+	var output []definitions.GettableApiReceiver
+	for i := q.Offset; i < len(rev.Config.AlertmanagerConfig.Receivers); i++ {
+		r := rev.Config.AlertmanagerConfig.Receivers[i]
+		if len(q.Names) > 0 && !slices.Contains(q.Names, r.Name) {
+			continue
+		}
+
+		// TODO: check for scoped read access and continue if not allowed
+
+		decryptAccess, err := rs.canDecrypt(ctx, user, r.Name)
+		if err != nil {
+			return nil, err
+		}
+		if q.Decrypt && !decryptAccess {
+			return nil, ErrPermissionDenied
+		}
+
+		decryptFn := func(uid string) func(v string) string {
+			return rs.decryptOrRedact(ctx, uid, decryptAccess && q.Decrypt)
+		}
+
+		res, err := rs.postableToGettableApiReceiver(r, provenances, decryptFn)
+		if err != nil {
+			return nil, err
+		}
+
+		// TODO: redact settings if the user only has list access
+
+		output = append(output, res)
+		// stop if we have reached the limit or we have found all the requested receivers
+		if (len(output) == q.Limit && q.Limit > 0) || (len(output) == len(q.Names)) {
+			break
+		}
+	}
+
+	return output, nil
+}
+
+func (rs *ReceiverGroupService) postableToGettableApiReceiver(r *definitions.PostableApiReceiver, provenances map[string]models.Provenance, decryptFn func(uid string) func(v string) string) (definitions.GettableApiReceiver, error) {
```

Comment:
I feel that this should be a function in a "compat.go" file. WDYT?

Reply:
Ideally this is the only place we make this conversion, but it doesn't hurt to pull them out.

Reply:
I believe there will be some in tests